### ISSUE TITLE
Pointwise layers are not converted tot int8 but to fp32/fp16 with TensorRT 8.4.2.4 when running on GPU RTX 3090

### ISSUE THREAD

**Issue by Michelvl92** [NVIDIA/TensorRT#3542]
## Description

I tried to convert a Yolov5s ONNX int8 model to a tensorRT int8 model. But all the PointWise operations are not converted to int8 as I would expect. By default, they are converted to FP32 precision, which is unexpected. This means that an "int8" model is higher than the inference latency of an FP16 model.

My goal is to reduce the inference latency of a yolov5s fp16 model, by quantizing it to int8, I expect that this can reduce the inference latency up to 2x. What can IO do to still increase the inference latency of the int8 model?

## Environment

I use the following Nvidia docker container: `nvcr.io/nvidia/pytorch:22.08-py3`
All the specs can be found in the [framework matrix here](https://docs.nvidia.com/deeplearning/frameworks/support-matrix/index.html)

**TensorRT Version**:

**NVIDIA GPU**: RTX 3090

**NVIDIA Driver Version**: 535.54.03

**TensorRT**: 8.4.2.4

**CUDA Version**: NVIDIA CUDA 11.7 Update 

**CUDNN Version**: 8.5.0.96

Operating System: Ubuntu 20.04

Python Version: python 3.8

PyTorch version: [1.13.0a0+d321be6](https://github.com/pytorch/pytorch/commit/d321be6)

Container: `nvcr.io/nvidia/pytorch:22.08-py3`

## Steps To Reproduce

1.  I have quantized a default pre-trained [yolov5s6](https://github.com/ultralytics/yolov5/tree/v6.2) model to int8 with Nvidia [PyTorch post-training quantization](https://docs.nvidia.com/deeplearning/tensorrt/pytorch-quantization-toolkit/docs/index.html)
2. I have exported the quantized model with [export.py](https://github.com/ultralytics/yolov5/blob/v6.2/export.py), which: converts the model to ONNX (optset 12), and then converts the model to TensorRT.
3. By inspecting the model TensorRT graph, I found out that the "int8" pointwise layers in the model are not converted to int8 but fp32. (see attached graph 1) The "fp16" model is faster than the "int8" model. 
4. I have added the option `config.set_flag(trt.BuilderFlag.FP16)`, and the pointwise layers in the model are now FP16. (see attached graph 2). Why are the pointwise layers not converted to int8 Precision? This makes the model far slower.

## Relevant Files
[Yolov5s6 "int8" model with int8 conversion options TRT graph](https://github.com/NVIDIA/TensorRT/assets/2651556/05a38523-95f5-4f96-9a80-ee9511033542)

[Yolov5s6 "int8" model with fp16/int8 conversion options TRT graph](https://github.com/NVIDIA/TensorRT/assets/2651556/05cd40a4-e9c4-4f07-afeb-20c689c0070e)

**Comment by Michelvl92** [comment]
I have tested it with a newer TensorRT version: 8.6.1, as installed in container: `nvcr.io/nvidia/pytorch:23.10-py3`, but I have the same problem. I have not tested it with TensorRT versions >= 9.x since this is not officially supported by: [trt-engine-explorer](https://github.com/NVIDIA/TensorRT/tree/main/tools/experimental/trt-engine-explorer)

**Comment by zerollzeng** [comment]
How did you place the Q/DQ here? Could you please share the onnx here?

One quick way to confirm is use the onnx without Q/DQ, and build the engine with `trtexec --onnx=model.onnx --int8 --fp16`, I think you should see it's running in INT8.

**Comment by Michelvl92** [comment]
@zerollzeng, [Nvidia Pytorch-Quantization-Toolkit](https://docs.nvidia.com/deeplearning/tensorrt/pytorch-quantization-toolkit/docs/index.html) is used to add Q/DQ. I used the Nvidia example here: [tutorials/quant_resnet50](https://docs.nvidia.com/deeplearning/tensorrt/pytorch-quantization-toolkit/docs/index.html#document-tutorials/quant_resnet50), by calling: _Automatic layer substitution is done with `quant_modules`. This should be called before model creation._

Would the onnx graph be enough info for you, here you can see the added Q/DQ nodes.
![exp8_best_yolov5s6_int8](https://github.com/NVIDIA/TensorRT/assets/2651556/35335538-62ff-4aed-91fc-5b10a6af5f1b)

**Comment by zerollzeng** [comment]
@ttyio should we add Q/DQ before the Mul layer?
![image](https://github.com/NVIDIA/TensorRT/assets/38289304/14c29e47-d9e5-4637-8e76-f71ab84feb93)

**Comment by nzmora-nvidia** [comment]
@Michelvl92 

I think you're missing two pairs of Q/DQ nodes - see the diagrams I've pasted and try adding the two pairs as shown (I've pasted the engine and ONNX graphs for the same subgraph). The Quant Toolkit only handles simple models and didn't add these.
![image](https://github.com/NVIDIA/TensorRT/assets/96238833/ecbc95d9-96cf-412e-995b-318edb990ef7)
Cheers
Neta

**Comment by Michelvl92** [comment]
@nzmora-nvidia Thank you for your answer, i already found this out myself, and I also added even more Q/DQ nodes as follows:

1. Q/DQ in both the inputs of the add in the bottleneck (as you showed) --> but it look likes it is failing, the output of the PWN(sigmoid, mul) is then converted to int8, why is the PWN(sigmoid, mul) not fused in the Conv? I think this is the cause, sometimes this happens, and sometimes not, see result onnx/tensorrt 
2. Q/DQ in the C3
3. Q/DQ all the inputs of the maxpool layer + one at the output in the SPPF block --> this worked!
4. Q/DQ in the input of the upsample layer --> this didn't failed, see onnx/tensorrt graph
5. Q/DQ in the other concat layers --> it looks like it worked?

Example  of a C3 block included bottleneck:
![yolov5_c3_arch](https://github.com/NVIDIA/TensorRT/assets/2651556/d46e6ce5-e64b-464e-a374-788cef277b37)

Example of a SPPF layer:
![sppf](https://github.com/NVIDIA/TensorRT/assets/2651556/f4b9a0e9-758b-413a-aa51-5ed1175d4f1a)

Q/DQ in both the inputs of the add in the bottleneck, but know the conv with pwn(sigmoid + mul) fails:
![onnx_quant_bottleneck](https://github.com/NVIDIA/TensorRT/assets/2651556/77d56b67-3e9d-452a-b4a3-f2611f0debfa)

![tensorrt_quant_bottleneck](https://github.com/NVIDIA/TensorRT/assets/2651556/ccaf301a-d267-4af0-9850-b1ab0e3b1035)

Fail to quantize resize (upsample) layer, it is correct in onnx, but not in tensorrt, why?
![tensorrt_quant_upsample](https://github.com/NVIDIA/TensorRT/assets/2651556/5c9c9d44-b4f0-4cf2-9c9b-4f79cfaaaba4)

![onnx_quant_upsample](https://github.com/NVIDIA/TensorRT/assets/2651556/4b3ae34a-1a3b-42a9-b1b1-a064518f634a)

The full onnx graph
![exp8_best_yolov5s6_int8_surgeon_v1](https://github.com/NVIDIA/TensorRT/assets/2651556/ac41847a-7321-457a-b611-c55f19927fc1)
![exp8_best_yolov5s6_int8_surgeon_lll engine graph json](https://github.com/NVIDIA/TensorRT/assets/2651556/0b880906-cb72-446d-ada2-b54c96e2d755)

**Comment by nzmora-nvidia** [comment]
Hi @Michelvl92, 
Please send the SVG images and ONNX files as attachments because it's too hard to follow this thread like this.
Are the results here from the latest TRT 8.6.x?

**Comment by Michelvl92** [comment]
@nzmora-nvidia, I have removed the large SVG/PNG graphs from the previous posts, and updated the question and problem.

The results are currently with TRT 8.4.2.4.

**Comment by Michelvl92** [comment]
@nzmora-nvidia, I have tried with TRT 8.6.1.6-1+cuda12.0.

See TRT graph:

1. resize is now in int8! --> This is fixed now with the new TRT version
2. For some strange reason conv + (sigmoid + mul) is still not always fused (in bottleneck layer), and the behaviour is very strange sometimes in the graph

[See TRT graph](https://github.com/NVIDIA/TensorRT/assets/2651556/9a27fcc1-8485-4526-a18f-4b51bbb1ecd3)

**Comment by nzmora-nvidia** [comment]
Thanks @Michelvl92, 
Indeed there are fusions that at face-value should occur and don't.
I can't find the attachment of the ONNX file itself so I can't debug - please attach.

I suspect something unexpected is going on with the scales of some of the QDQ nodes.
Can you verify that the scales of "/model/model/model.2/m/m.0/cv1/conv/_input_quantizer/QuantizeLinear" and "/model/model/model.2/m/m.0/addop/_input0_quantizer/QuantizeLinear" are the same?

BTW, TRT graphs in SVG format have more information in them (i.e. pop-up when hovering over an operator).

**Comment by Michelvl92** [comment]
@nzmora-nvidia thanks for your replay, with onnx-surgeon I have removed and reconnected the double Q/DQ nodes, and it looks now correct as you can see. Almost all layers are now int8, only not the last ones at the end, which is not a big problem.

Do you see any points to optimize? The total model speedup of the int8 model compared to the FP16 model is only 14%, is this what you would expect? I would have expected a speedup of 40%.
- The input can be made int8 since the reformat in the input takes up to 10% of the total latency budget
- The speed of the max-pooling is similar as in the fp16 model, why?
- The speed of the resize is similar as in the fp16 model, why?
- For both max-pooling and resizing there are scale operations added that take 3.% of the latency budget, why?
- Why is the pointwise layer at some parts not fused?

**Do you see how I can make better speed improvements?**

Thanks!

[TRT graph](https://github.com/NVIDIA/TensorRT/assets/2651556/f656e972-e360-4b72-b5ea-cdc6cdaae6e7)

**Comment by Raj-vivid** [comment]
> [@Michelvl92](https://github.com/Michelvl92) w/o the ONNX I cannot debug so I'm letting you know the problems that I see in the output engine, but I cannot be sure of their root-cause.
> 
> * Check if `/model/model/model.33/detectquantinput/QuantizeLinear` and `/model/model/model.24/conv/_input_quantizer/QuantizeLinear` have the same scale value. If they don't that would explain why they don't fuse.
> * The extra scale ops are unfused Q/DQ operators (e.g. the two Q operators in the bullet above and `/model/model/model.33/detectquantinput/DequantizeLinear`)
> * `/model/model/model.33/m.0/Conv` looks problematic. Check if you quantized it correctly.
> * These 2 issues repeat several times at the lower part of the graph and that's why you see some no quantized layers.
> * Quantizing MaxPooling and Resize for small shapes (like yours) may not help much (GPU utilization) but you get the benefit of Conv outputs being quantized so there's overall benefit.

Could you give me some general pointers for how you handle explicit fp32 operations present in the graph? I am using netron to visualize the quantized graph and I see a lot of mul, add, pow, transpose, mean_reduce operations in fp32.
